In [213]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

In [214]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [215]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [216]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=1200,
    height=700,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [217]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
#from df_market_data we are using the standard scaler to ensure a zero mean so that we have consistency between the variables
#the fit transform then calculares the mean as well as the SD
scaled_data = StandardScaler().fit_transform(df_market_data)

In [218]:
# Create a DataFrame with the scaled data
#The purpose of this line of code is that it contins the scaled data (from the line above) as well as the normailzed version of the original data 
scaled_market_data = pd.DataFrame(scaled_data,columns=df_market_data.columns)

# Copy the crypto names from the original data
# In this line we are adding the colmn coin_id from the original data
scaled_market_data["coin_id"] = df_market_data.index.tolist()

# Set the coinid column as index
scaled_market_data = scaled_market_data.set_index("coin_id")

# Display sample data
scaled_market_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [219]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [220]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list 

for k in range(1, 12):
    kmeans = KMeans(n_clusters=k)
    kmeans.fit(scaled_market_data)
    inertia_values.append(kmeans.inertia_)

In [221]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k': range(1, 12),'inertia': inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow

k     inertia
0    1  287.000000
1    2  198.571818
2    3  123.190482
3    4   79.022435
4    5   63.858668
5    6   52.592907
6    7   43.989781
7    8   36.791070
8    9   32.719192
9   10   28.369450
10  11   24.048292

In [222]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot = df_elbow.hvplot.line(x="k", y="inertia", title="Elbow Curve")
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Using the elbow method it looks as if 4 is the best as the change is not rapid anymore.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [223]:
# Initialize the K-Means model using the best value for k
kmeans = KMeans(n_clusters=4)

In [224]:
# Fit the K-Means model using the scaled data
kmeans.fit(df_market_data_scaled)

KMeans(n_clusters=4)

In [225]:
# Predict the clusters to group the cryptocurrencies using the scaled data
crypto_clusters = kmeans.predict(scaled_market_data)

# Print the resulting array of cluster values.
print(crypto_clusters)

[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 2 1 0 0 3
 0 0 0 0]


In [226]:
# Create a copy of the DataFrame
scaled_predictions = scaled_market_data.copy()

In [227]:
# Add a new column to the DataFrame with the predicted clusters
scaled_predictions["crypto_cluster"] = crypto_clusters

# Display sample data
scaled_predictions

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [228]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scatter_plot = scaled_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    hover_cols=["coin_id"],
    by="crypto_cluster",
    marker=["hex", "square", "cross", "inverted_triangle"],
    title="Cryptocurrencies Clusters"
)
scatter_plot

:NdOverlay   [crypto_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [229]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [230]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_data = pca.fit_transform(scaled_market_data)

# View the first five rows of the DataFrame.
pca_data[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [231]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [232]:
total_explained_variance = sum(pca.explained_variance_ratio_)
total_explained_variance

0.8950316570309842

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 89.5%, which is high as it means that three principal components retained a large portion of the original data

In [233]:
# Creating a DataFrame with the PCA data
pca_data = pd.DataFrame(
    market_pca_data,
    columns=["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original data
pca_data["coin_id"] = df_market_data.index


# Set the coinid column as index
pca_data = pca_data.set_index("coin_id")

# Display sample data
pca_data

PC1       PC2       PC3
coin_id                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 -0.732719 -0.062543
cosmos               -0.306272  0.703415  1.714224
cdai                 -0.513528 -0.142802 -0.656566
neo                  -0.362120 -0.986914 -0.728752
wrapped-bitcoin      -0.604265  0.827398  0.439316
leo-token            -0.413296 -0.674115 -1.076628
huobi-token          -0.407483 -0.212507 -0.351426
nem                   0.608974  0.563532 -1.148742
binance-usd          -0.450211 -0.151019 -0.647401
iota                 -0.764665 -0.517886  0.204990
vechain              -0.556315 -1.938209 -1.261776
zcash                -0.425147  0.492976  1.058048
theta-token           2.676868 -0.013954 -1.965207
dash                 -0.613923 -0.479337  0.339565
ethereum-classic     -0.579924 -0.356334 -0.114942
ethlend               8.089018 -3.896891  2.301382
maker                -0.389045  0.165041  0.379414
havven                0.865762 -2.261882  0.275583
omisego               0.111675  0.428316 -1.205398
celsius-degree-token  4.792395  6.767679 -1.986985
ontology             -0.632355 -2.108117 -0.652227
ftx-token            -0.593142  0.021485  0.209911
true-usd             -0.458131 -0.135734 -0.635284
digibyte             -0.297910 -0.191126 -0.909602

---

### Find the Best Value for k Using the PCA Data

In [234]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [235]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in range(1, 12):
    model = KMeans(n_clusters=k)
    model.fit(pca_data)
    inertia.append(model.inertia_)

In [236]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow = {"k": range(1, 12), "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_df = pd.DataFrame(pca_elbow)
df_elbow_df

k     inertia
0    1  256.874086
1    2  165.901994
2    3   93.774626
3    4   49.665497
4    5   37.878747
5    6   27.618972
6    7   21.182776
7    8   17.016962
8    9   13.814082
9   10   10.543635
10  11    8.058822

In [237]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot_pca = df_elbow_df.hvplot.line(x="k", y="inertia", title="Elbow Curve using PCA Data")

elbow_plot_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best k value would be 4 again as the change isnt as rapid 
  



* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Its similar to k value from the original data, looks ot be almost identical

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [238]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters = 4)

In [239]:
# Fit the K-Means model using the PCA data
model_pca.fit(pca_data)

KMeans(n_clusters=4)

In [240]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_clusters = model_pca.predict(pca_data)
# Print the resulting array of cluster values.
print(crypto_clusters_pca)

[0 0 3 3 0 0 0 0 0 3 3 3 3 0 3 0 3 3 0 3 3 0 3 3 3 3 3 3 0 3 3 3 1 0 3 3 2
 3 3 3 3]


In [241]:
# Create a copy of the DataFrame with the PCA data
df_market_data_pca_predictions = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_data_pca_predictions["crypto_cluster"] = crypto_clusters_pca

# Display sample data
df_market_data_pca_predictions

PC1       PC2       PC3  crypto_cluster
coin_id                                                           
bitcoin              -0.600667  0.842760  0.461595               0
ethereum             -0.458261  0.458466  0.952877               0
tether               -0.433070 -0.168126 -0.641752               3
ripple               -0.471835 -0.222660 -0.479053               3
bitcoin-cash         -1.157800  2.041209  1.859715               0
binancecoin          -0.516534  1.388377  0.804071               0
chainlink            -0.450711  0.517699  2.846143               0
cardano              -0.345600  0.729439  1.478013               0
litecoin             -0.649468  0.432165  0.600303               0
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653               3
crypto-com-chain     -0.248198 -1.376252 -1.462026               3
usd-coin             -0.438408 -0.175337 -0.663388               3
eos                  -0.693425 -0.473815 -0.527597               3
monero                0.060499  2.909404  1.498571               0
tron                 -0.393352 -0.108192 -0.012756               3
tezos                -0.796176 -0.494409  1.082812               0
okb                   0.064075 -1.269825 -1.098829               3
stellar              -0.489015 -0.732719 -0.062543               3
cosmos               -0.306272  0.703415  1.714224               0
cdai                 -0.513528 -0.142802 -0.656566               3
neo                  -0.362120 -0.986914 -0.728752               3
wrapped-bitcoin      -0.604265  0.827398  0.439316               0
leo-token            -0.413296 -0.674115 -1.076628               3
huobi-token          -0.407483 -0.212507 -0.351426               3
nem                   0.608974  0.563532 -1.148742               3
binance-usd          -0.450211 -0.151019 -0.647401               3
iota                 -0.764665 -0.517886  0.204990               3
vechain              -0.556315 -1.938209 -1.261776               3
zcash                -0.425147  0.492976  1.058048               0
theta-token           2.676868 -0.013954 -1.965207               3
dash                 -0.613923 -0.479337  0.339565               3
ethereum-classic     -0.579924 -0.356334 -0.114942               3
ethlend               8.089018 -3.896891  2.301382               1
maker                -0.389045  0.165041  0.379414               0
havven                0.865762 -2.261882  0.275583               3
omisego               0.111675  0.428316 -1.205398               3
celsius-degree-token  4.792395  6.767679 -1.986985               2
ontology             -0.632355 -2.108117 -0.652227               3
ftx-token            -0.593142  0.021485  0.209911               3
true-usd             -0.458131 -0.135734 -0.635284               3
digibyte             -0.297910 -0.191126 -0.909602               3

In [242]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
scatter_plot_pca = df_market_data_pca_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    hover_cols=["coin_id"],
    by="crypto_cluster",
    marker=["hex", "square", "cross", "inverted_triangle"],
    title="Cryptocurrencies Clusters PCA"
)
scatter_plot_pca

:NdOverlay   [crypto_cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [243]:
# Composite plot to contrast the Elbow curves
elbow_plot + elbow_plot_pca

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [244]:
# Composite plot to contrast the clusters
scatter_plot + scatter_plot_pca

:Layout
   .NdOverlay.I  :NdOverlay   [crypto_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [crypto_cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The elbow charts look to be almost identical but still look to have a valid k value at 4. As for the clusters I thought my clusters were inverted but after further analysis I see that it looks to have two outliers within it being in 2 and 3 in the first and 1 and 2 in the second. In the first clustering there was a negative on the far left side but in the second cluster the outlier is on the far right. Althogh there are outliers in both with two clusters forming around 0 we can clearly see that the cluster is tigher in the second. 